### PR TITLE
[Omega] Extract default kid from init data if required

### PR DESCRIPTION
**Pull request by matthuisman**
## Description
Sending the below as license_data property
`AAAAMnBzc2gAAAAA7e+LqXnWSs6jyCfc1R0h7QAAABISEByIsi1reEZguaxNxHvMJt8=`

Kodi 20 - OK
```
2024-11-10 20:09:10.652 T:1943  debug <general>: AddOnLog: inputstream.adaptive: Searching PSSH data in FILE
2024-11-10 20:09:10.652 T:1943  debug <general>: AddOnLog: inputstream.adaptive: Initializing stream with KID: 1c88b22d6b784660b9ac4dc47bcc26df
```

Kodi 21 - ERROR
```
2024-11-10 11:42:10.723 T:1003 debug <general>: AddOnLog: inputstream.adaptive: License data: Use PSSH data provided by the license data property
2024-11-10 11:42:10.723 T:1003 error <general>: AddOnLog: inputstream.adaptive: ConvertKidStrToBytes: Cannot convert KID "" as bytes due to wrong size
2024-11-10 11:42:10.723 T:1003 warning <general>: AddOnLog: inputstream.adaptive: Initializing stream with unknown KID!
```

Kodi 22 - OK
```
2024-11-10 21:24:47.437 T:22820   debug <general>: AddOnLog: inputstream.adaptive: License data: Use PSSH data provided by the license data property
2024-11-10 21:24:47.437 T:22820   debug <general>: AddOnLog: inputstream.adaptive: Default KID parsed from init data
2024-11-10 21:24:47.438 T:22820   debug <general>: AddOnLog: inputstream.adaptive: Initializing stream with KID: 1c88b22d6b784660b9ac4dc47bcc26df
```

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/Session.cpp [hunk 1]
```diff
@@ -469,6 +469,17 @@ bool CSession::InitializeDRM(bool addDefaultKID /* = false */)
         initData = BASE64::Decode(licenseDataStr);
       }
 
+      // If no KID, but init data, extract the KID from init data
+      if (!initData.empty() && defaultKidStr.empty())
+      {
+        DRM::PSSH parser;
```
Comment (matthuisman) [inline src/Session.cpp]: oh, this might not even be available in Omega

File: src/Session.cpp [hunk 2]
```diff
@@ -441,7 +442,7 @@ bool CSession::InitializeDRM(bool addDefaultKID /* = false */)
                      licenseData.empty() ? "" : "(with custom data)");
 
             std::vector<uint8_t> wvPsshData;
-            if (DRM::MakeWidevinePsshData(defaultKid, licenseData, wvPsshData))
+            if (DRM::MakeWidevinePsshData(DRM::ConvertKidStrToBytes(defaultKidStr), licenseData, wvPsshData))
```
Comment (matthuisman) [inline src/Session.cpp:445]: not sure if need { } around DRM::ConvertKidStrToBytes(defaultKidStr)

**Comment by matthuisman** [comment]
this isnt actually working, I still get
```
2024-11-10 21:44:44.730 T:21220   debug <general>: AddOnLog: inputstream.adaptive: License data: Use PSSH data provided by the license data property
2024-11-10 21:44:44.730 T:21220   error <general>: AddOnLog: inputstream.adaptive: DRM::ConvertKidStrToBytes: Cannot convert KID "" as bytes due to wrong size
2024-11-10 21:44:44.730 T:21220 warning <general>: AddOnLog: inputstream.adaptive: Initializing stream with unknown KID!
```

**Comment by CastagnaIT** [comment]
this because your PSSH data is wrong, PSSH version 0 dont support kids container

**Comment by matthuisman** [comment]
oh, thats because the parser in kodi 21 doesnt GetKeyIds for version 0 from the data..
Trying again now with version 1 with same keyid in kids...

**Comment by matthuisman** [comment]
so problems with Kodi 21:
1) does not create default kid from initdata (Kodi 20 and 22 both do)
2) does not support version 0 getting KID from data (Kodi 20 and 22 both do)

**Comment by CastagnaIT** [comment]
> does not support version 0 getting KID from data (Kodi 20 and 22 both do)

ah yeah now remember,
on 22 i reworked the pssh parser and all others related code, and that's why i didn't backport,
too much code to change and it wasn't worth investing time with the risk of regressions

so on 22 the parser is able to get kids from widevine data, contained in the PSSH data (regardless of PSSH version)

**Comment by matthuisman** [comment]
when I use
AAAARnBzc2gBAAAA7e+LqXnWSs6jyCfc1R0h7QAAAAEciLIta3hGYLmsTcR7zCbfAAAAEhIQHIiyLWt4RmC5rE3Ee8wm3w==

it gets further
```
2024-11-10 22:03:36.659 T:20768   debug <general>: AddOnLog: inputstream.adaptive: License data: Use PSSH data provided by the license data property
2024-11-10 22:03:36.659 T:20768   debug <general>: AddOnLog: inputstream.adaptive: Default KID parsed from init data
2024-11-10 22:03:36.659 T:20768   debug <general>: AddOnLog: inputstream.adaptive: Initializing stream with KID: 
2024-11-10 22:03:36.694 T:20768   debug <general>: AddOnLog: inputstream.adaptive: CDMMessage: 1 arrived!
```

it plays regardless of this issue, so im not sure if the default kid is being set correctly.
I know a user has issues on Linux64 with this stream so thinking this is the problem there.

"Initializing stream with KID: "
Its just showing blank...
Maybe thats just a type issue? cant convert to str ?

**Comment by matthuisman** [comment]
On 22, the parser can get kid from the DATA only (version 0) :)
here is where it does that: https://github.com/xbmc/inputstream.adaptive/blob/Piers/src/decrypters/Helpers.cpp#L353

Its not a biggy, I can add the key as well.
But would still need the fix to grab default_kid from init_data (this PR)

**Comment by matthuisman** [comment]
oh, I see

LOG::Log(LOGDEBUG, "Initializing stream with KID: %s", sessionPsshset.defaultKID_.c_str());

We logging sessionPsshset.defaultKID not defaultkid..

**Comment by matthuisman** [comment]
Fixed that with latest commit :)

**Comment by matthuisman** [comment]
success, with this PR and adding kids to my license_data:
AAAARnBzc2gBAAAA7e+LqXnWSs6jyCfc1R0h7QAAAAEciLIta3hGYLmsTcR7zCbfAAAAEhIQHIiyLWt4RmC5rE3Ee8wm3w==

```
2024-11-10 22:17:26.709 T:22348   debug <general>: AddOnLog: inputstream.adaptive: media::CdmAdapter::OnInitialized: CDM is initialized: true
2024-11-10 22:17:26.709 T:22348   debug <general>: AddOnLog: inputstream.adaptive: License data: Use PSSH data provided by the license data property
2024-11-10 22:17:26.709 T:22348   debug <general>: AddOnLog: inputstream.adaptive: Default KID parsed from init data
2024-11-10 22:17:26.709 T:22348   debug <general>: AddOnLog: inputstream.adaptive: Initializing stream with KID: 1c88b22d6b784660b9ac4dc47bcc26df
2024-11-10 22:17:26.743 T:22348   debug <general>: AddOnLog: inputstream.adaptive: CDMMessage: 1 arrived!
```

**Comment by CastagnaIT** [comment]
seem there is a bit of mess with vars (also on 22)

e.g. ExtractStreamProtectionData set extracted kid to sessionPsshset but dont update defaultKidStr

also on condition below
`if (sessionPsshset.pssh_ == m_adaptiveTree->m_currentPeriod->GetPSSHSets()[i].pssh_)`
imo `sessionPsshset.pssh_` should be updated by `initData` when needed

I've always found this part confusing, although with the rework i will delete all this

**Comment by matthuisman** [comment]
Understood. But this PR is not a rework. It's fixing a regression so I'm trying to keep changes minimal. Is this able to go in to fix the issue? Otherwise I'll need to proxy manifest and add in content protection section

**Review by CastagnaIT — APPROVED** [review]
its ok,
just please add prefix to commit title otherwise will make more hard to read the git history
in your case you are modify the session, so `[Session] xy...`

**Comment by matthuisman** [comment]
no worries :) I have updated the commit message

**Comment by matthuisman** [comment]
Thank you @CastagnaIT  :)

**Comment by matthuisman** [comment]
@CastagnaIT  if there are no more planned changes coming soon for Omega - are you able to do a release with this change ?

**Comment by CastagnaIT** [comment]
yes the intention is to make releases by saturday

**Pull request merged**